File: backtracking/rust/src/all_permutations.rs
use std::collections::HashMap;
pub fn all_permutations(n: i32, k: i32) {
    let mut current_sequence: Vec<i32> = Vec::new();
    let mut map: HashMap<i32, bool> = HashMap::new();
    for i in 1..n + 1 {
        map.insert(i, false);
    }

    all_permutations_help(&mut current_sequence, n, k, &mut map);
}

fn all_permutations_help(tmp: &mut Vec<i32>, n: i32, k: i32, used: &mut HashMap<i32, bool>) {
    if tmp.len() as i32 == k {
        println!("{:?}", tmp);
    }
    for i in 1..n + 1 {
        let v = used.get(&i);
        match v {
            None => panic!("system failed"),
            Some(x) => {
                if *x {
                    continue;
                } else {
                    used.insert(i, true);
                    tmp.push(i);
                    all_permutations_help(tmp, n, k, used);
                    tmp.pop();
                    used.insert(i, false);
                }
            }
        }
    }
}

File: backtracking/python/all_permutations.py
"""
    获取n个树中取k个的全部排列
"""


def all_permutations(n: int, k: int) -> None:
    if n < k:
        raise("无效参数")
    used = [0 for i in range(n)]
    all_permutations_help(n, k, [], used)


def all_permutations_help(n: int, k: int, tmp: list[int], used: list[int]):
    if len(tmp) == k:
        print(tmp)
        return

    for i in range(n):
        if used[i] == 0:
            tmp.append(i+1)
            used[i] = 1
            all_permutations_help(n, k, tmp, used)
            used[i] = 0
            tmp.pop()


if __name__ == "__main__":
    all_permutations(4, 2)
